<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/html4/loose.dtd">
<html><!-- InstanceBegin template="/Templates/standard.dwt" codeOutsideHTMLIsLocked="false" -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=">
<!-- InstanceBeginEditable name="doctitle" -->
<title>Tutorial</title>
<!-- InstanceEndEditable -->
<link rel="stylesheet" href="styles/mesquitePackage.css" type="text/css">
<!-- InstanceBeginEditable name="head" --><!-- InstanceEndEditable -->
</head>


<body>
<div id="masthead">
  <div id="banner"><img src="splash.gif" width="164" height="64"></div>
  <div id="mainTitle"><font id="packageName"><b>Chromaseq</b></font><BR>
  <font id="version">Test Version 0.973<br>
  8 January 2010</font><BR>
         David R. Maddison and Wayne P. Maddison</div>
  <div id="subtitle"><!-- InstanceBeginEditable name="subtitle" -->
    <p>Tutorial</p>
    <!-- InstanceEndEditable --></div>
 
</div>
<div>
<!-- end masthead -->
<div id="content"><!-- InstanceBeginEditable name="content" -->
  <p>In this tutorial,
      you will go from chromatograms to having fully processed sequences, with
      final base calls.</p>
  <h2>Setup</h2>
  <p>First, make sure that you the Chromaseq package installed, and that you
    have <a href="installingPP.html">installed
      Phred and Phrap</a>.</p>
  <p>Download the <a href="http://mesquiteproject.org/packages/chromaseq/examples/ChromaseqExample.zip">ChromaseqExamples archive</a>,
    and unzip it. You will have a folder containing four things:</p>
  <ul>
    <li>a folder containing some example
            chromatograms</li>
    <li>a file called &quot;PrimerList&quot; containing the <a href="primers.html">names
                of primers</a></li>
    <li>a file called &quot;SampleCodes&quot; showing the correspondence between <a href="sampleCode.html">sample
                codes and sample names</a></li>
    <li>a  file called &quot;PrimerSequences&quot;
      containing <a href="primerSequences.html">primer
                sequences</a>.</li>
    </ul>
  <p>You will also need to modify the phredpar.dat file that come with Phred,
    so that it can accommodate the chromatograms provided in the example. Open
    up the phredpar.dat file in a simple text editor, and </p>
  <p><font face="Courier New, Courier, mono">  &nbsp;&nbsp;&quot;BeckmanCEQ&quot; terminator
    big-dye	Beckman_CEQ_2000</font></p>
  <p>and immediately below this add the following two lines:</p>
  <p><font face="Courier New, Courier, mono">&nbsp;&nbsp;#Entry below added to accommodate Chromaseq example files<br>
&nbsp;&nbsp;&quot;DT3730POP7{BDv3}.mob&quot; terminator big-dye ABI_373_377</font></p>
  <p>From the File&gt;Open Other submenu in Mesquite, choose &quot;Process Chromatogram Files in Directory&quot;. You will see a dialog box that looks something
    like this:</p>
  <p align="center"><img src="images/mainPhPhDialogEmpty.gif" width="358" height="437"></p>
  <p>&nbsp;</p>
  <p>Let's begin at the bottom, by telling Mesquite the locations of Phred, Phrap,
    and various other files. </p>
  <p>Press on the
      &quot;Phred Phrap Locations &amp; Options...&quot; button, and the dialog will appear
    that will
      let
      you tell Mesquite
      where
      the programs are located, and where the Phred parameter file is:</p>
  <p align="center"><img src="images/PhPhLocations.gif" width="503" height="200"></p>
  <p>Press the &quot;Browse...&quot; button beside the &quot;Phred, Phrap, &amp; Phd2Fasta&quot; to browse
    for the folder on your hard disk containing the three programs Chromaseq
    uses
    (Phred,
    Phrap,
    and
    Phd2Fasta),
    all of which must be in the same directory. Do the same for the Phred parameter
    file (&quot;phredpar.dat&quot;). You should only have to do this once on your computer,
    unless you move these files to other locations. </p>
  <p>Once that is done, then you will need to tell Chromaseq three more things
    to set up the system:</p>
  <ul>
    <li> How to find, in the file
      names of the chromatograms, the sample codes and primer names. To do this
      for your own sequences, you would need to create a <a href="namingRules.html">naming
      rule</a>. However, for these sample chromatogram files, the default naming
      rule will work.</li>
    <li><em> Where the file is that lists the sample codes and the
      names of the samples.</em> Use the &quot;Browse&quot; button beside the Code &amp; Names File
      area and choose the file &quot;SampleCodes&quot; from the download.</li>
    <li> <em>Where the file is that lists the names of the primers, whether they
      are forward or reverse primers, and the name of gene or genomic piece to
      which the belong.</em> Use the &quot;Browse&quot; button beside the &quot;Primer
      list file&quot; area and choose the file &quot;PrimerList&quot; from the download.</li>
    </ul>
  <p>When you use this system to process your own sequences, you would need to
    provide your own naming rule, sample code file, and primer list. But, once
    these are set up, you will in general not need to re-enter these. </p>
  <h2>Asking Phred and Phrap to make base calls and assemble contigs</h2>
  <p>With the set-up completed, you can now use the File&gt;Open Other&gt;Phred/Phrap
    Import from Local Chromatograms dialog box to ask Phred and Phrap to process
    your files. Press the &quot;OK&quot; button, and Chromaseq will ask you which directory
    contains the chromatograms. Choose the &quot;Chromatograms&quot; folder from the
    download. </p>
  <p>At this point, your computer's terminal shell should start. On a Mac, this
    will typically be the Terminal application; on Windows, it will be the Cygwin
    shell if that is what you installed. Phred and Phrap will then start processing
    the files, and the shell will be filled with the output of these programs.
    If you switch back to Mesquite, you will notice in the log information about
    what is happening. Eventually, the processing will finish, and Mesquite will
    import the resulting contigs into a Mesquite project, with one matrix for
    each gene.</p>
  <p>In the example in this tutorial, the chromatograms come from three different
    genes, and so three matrices will appear in the project once the processing
    is done, one for 18S rDNA, one for COI, and the other for the gene <em>wingless</em>    (called
    &quot;wg&quot; in the PrimerList file). The last matrix processed will be presented
    to you, by default in Birds Eye View, and with the bases colored according
    to the quality score assigned to them by Phred and Phrap. White is high quality,
    darker greens are moderate quality, and blue is poor quality:</p>
  <p align="center"><img src="images/tutorialImport.gif" width="828" height="335"></p>
  <p>The pink sequence means the contig was not successfully assembled.</p>
  <h2>Examining a contig and its chromatograms</h2>
  <p>To see the contig for the wg sequence of Bemb antiquum IA 1626 and its chromatograms,
    touch the view chromatograms tool (<img src="images/chromTool.jpg" width="30" height="29" align="texttop">)
    on &quot;Bemb antiquum IA 1626&quot;. A new window will open, showing you the chromatograms
    and the contig:</p>
  <p align="center"><img src="images/tutorialChromato1.gif" width="828" height="395"></p>
  <p>You may see a different part of the sequence than is shown in the above
    image; Chromaseq centers the view on the base upon which you clicked in the
    matrix.</p>
  <p align="center"><img src="images/tutorialFirstChromView.gif" width="780" height="331"></p>
  <p>&nbsp;</p>
<!-- InstanceEndEditable --></div>
<!--end content -->
<div id="navBar">
  <div id="sectionLinks">
    <ul>
      <li><a href="index.html">Home</a></li>
      <li><a href="overview.html">Overview</a></li>
      <li><a href="quickstart.html">Quickstart</a></li>
	  <li><a href="tutorial.html">Tutorial</a></li>
	  <li><a href="phredphrap.html">Phred / Phrap with Chromaseq</a></li>
	  <ul id="subUL">
	  	<li><a href="installingPP.html">Installing Phred &amp; Phrap </a></li>
  		<li><a href="primers.html">Primer List File</a></li>
  		<li><a href="sampleCode.html">Sample Codes &amp; Names File</a></li>
  		<li></li>
  		<li><a href="runningPP.html">Running Phred/Phrap</a></li>
	  	<li><a href="namingRules.html">Chromatogram File Naming Rules </a></li>
		<li><a href="quality.html">Quality Scores</a></li>
		<li><a href="postPhrap.html">Post-Phrap Options</a></li>
	  </ul>
      <li><a href="viewer.html">Chromatogram Viewer</a></li>
  		<ul id="subUL">
	  		<li><a href="navigation.html">Navigation</a></li>
  			<li><a href="display.html">Display Options</a></li>
  			<li><a href="callingBases.html">Calling Bases</a></li>
		</ul>
      <li><a href="genbank.html">GenBank</a></li>
      <li><a href="otherPrograms.html">Sequencher &amp; Others </a></li>
      <li><a href="acknowledgements.html">Acknowledgements</a></li>
      <li><a href="citation.html">Citation</a></li>
      <li><a href="faq.html">FAQs</a></li>
      <li><a href="future.html">Future Directions</a></li>
      <li style="margin-top: 20px;"><a href="download.html">Download</a></li>
      <li><a href="sourceCode.html">Source Code</a></li>
    </ul>
	<p>&nbsp;</p>
	<ul><li> <a href="http://mesquiteproject.org"  style="border: 0px;"><img src="images/mesquiteIcon.gif" alt="Mesquite Home" width="32" height="32"> Mesquite Home</a> </li>
    </ul>
  </div>
</div>
</div>
<div id="navBarExtension">&nbsp;</div>
<!--end navbar -->
<div id="siteInfo"> Copyright &copy; 2010 <a href="http://david.bembidion.org">David
    R. Maddison</a> and <a href="http://salticidae.org/wpm/home.html">Wayne P.
    Maddison</a> </div>
<br>
</body>
<!-- InstanceEnd --></html>
